src/stacking.lisp

```diff
@@ -18,59 +40,56 @@ This function should be used to implement copying of contiguous row-major blocks
-    (copy-row-major-block source destination element-type
-                          :destination-start (* start-row (ncol destination)))))
+;; TODO: Are these the best names for this and 'cols', following)?
+(defun rows (object)
+  "Return the number of rows of an array or scalar. Scalars return 1."
+  (etypecase object
+    (array (ecase (rank object)
+             (1 1)
+             (2 (dim object 0))))
+    (number 1)))
+
+;; TODO: Are these the best names for this and 'rows', previous)?
+(defun cols (object)
+  "Return the number of columns of an array or scalar. Scalars return 1."
+  (etypecase object
+    (array (ecase (rank object)
+             (1 (dim object 0))
+             (2 (dim object 1))))
+    (number 1)))
 
 (defun stack-rows* (element-type &rest objects)
-  "Stack OBJECTS row-wise into an array of the given ELEMENT-TYPE, coercing if necessary.  Always return a simple array of rank 2.
-
-How objects are used depends on their dimensions, queried by DIMS:
-
-- when the object has 0 dimensions, fill a row with the element.
-
-- when the object has 1 dimension, use it as a row.
-
-- when the object has 2 dimensions, use it as a matrix.
-
-When applicable, compatibility of dimensions is checked, and the result is used to determine the number of columns.  When all objects have 0 dimensions, the result has one column."
-  (let (ncol)
-    (flet ((check-ncol (dim)
-             (if ncol
-                 (assert (= ncol dim))
-                 (setf ncol dim))))
-      (let* ((nrow 0)
-             (start-rows-and-dims (mapcar
-                                   (lambda (object)
-                                     (let* ((dims (dims object))
-                                            (increment (ematch dims
-                                                         (nil 1)
-                                                         ((list d0) (check-ncol d0)
-                                                          1)
-                                                         ((list d0 d1) (check-ncol d1)
-                                                          d0))))
-                                       (prog1 (cons nrow dims)
-                                         (incf nrow increment))))
-                                   objects))
-             (ncol (or ncol 1)))
-        (let ((result (make-array (list nrow ncol) :element-type element-type)))
-          (mapc (lambda (start-rows-and-dims object)
-                  (destructuring-bind (start-row &rest dims)
-                      start-rows-and-dims
-                    (if dims
-                        (stack-rows-copy object result element-type start-row)
-                        (fill (displace result ncol (* start-row ncol))
-                              (coerce object element-type)))))
-                start-rows-and-dims objects)
-          result)))))
+  "Stack OBJECTS, row-wise, into a simple 2D array of given ELEMENT-TYPE.
```

Comment:
Thanks! This is much clearer to me as well. I think I actually understand what your code is doing...

Reply:
Glad to hear it. :-)

Like nearly every thing re-written, I started out sure I could make it very clean and concise.
Then, I realized all the corner cases, nuances, etc. :-)

Reply:
As long as we have the unit tests to catch the corner cases etc. then all should be well. Thanks for this!